Microsoft.DotNet.Interactive/Events/InputRequested.cs

```diff
@@ -0,0 +1,51 @@
+// Copyright (c) .NET Foundation and contributors. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using Microsoft.DotNet.Interactive.Commands;
+using Newtonsoft.Json;
+
+namespace Microsoft.DotNet.Interactive.Events
+{
+    public class InputRequested : KernelEventBase
+    {
+        [JsonConstructor]
+        internal InputRequested(string prompt, bool isPassword)
+        {
+            Prompt = prompt;
+            IsPassword = isPassword;
+        }
+
+        public InputRequested(
+            string prompt,
+            bool isPassword,
+            IKernelCommand command) : base(command)
+        {
+            if (command == null)
+            {
+                throw new ArgumentNullException(nameof(command));
+            }
+
+            Prompt = prompt;
+            IsPassword = isPassword;
+        }
+
+        [JsonIgnore]
+        private string _content;
+
+        [JsonIgnore]
+        public string Content
+        {
+            get => AllowStdin ? _content : throw new NotSupportedException("The stdin channel is not allowed by the frontend.");
```

Comment:
Throwing on a getter can be a gotcha. What's the scenario for allowing this event to be created at all in the case where a stdin channel is not supported? Can we handle this higher up?

Reply:
Hmm, yes, handling this higher up would be better. Let me play this a bit more.

Reply:
@jonsequitur Could you please take another look? I add the `AllowStdin` property to `SubmitCode`, and set that in `ExecuteRequestHandler.Handle` when constructing the `SubmitCode` instance.

Reply:
I suspect that the decision to allow stdin would apply to the whole session, based on the frontend, and therefore be more a configuration-level concern than something that would be communicated at the command level. 

We've been thinking for a while about introducing some API to learn more about the client since different clients have different capabilities. This seems related.

Reply:
> I suspect that the decision to allow stdin would apply to the whole session, based on the frontend

I also feel this way. But the fact that the `allow_stdin` is passed along by `execute_request` message makes me wonder if it could be possible a client allows the stdin channel sometimes but not others.

Reply:
Interesting. While that's possible, I prefer to avoid letting Jupyter drive the design of the core abstractions. 

Maybe the Jupyter layer could modify this behavior via middleware?

Reply:
A few tests are added to the Jupyter layer to validate the `input_request` message is issued correctly.

Capture our conversation here:
> One approach might be in the `input` helper method, check the current frontend's capabilities, and make a decision there.
Checking the capability of the frontend is something we could access from the `KernelInvocationContext` potentially.
It might be a property bag of sorts, with strong-typed access.
Air code:
>```
>if (!KernelInvocationContext.Current.Get<ClientConfiguration>().IsStdInSupported)
>{
>    await KernelInvocationContext.Current.DisplayErrorAsync("Sorry!");
>}
>```
>No protocol message to check for the frontend capability,
>but there are behavioral differences between frontends that we've considered sniffing somehow in order to have a good experience "out of the box" on Jupyter Lab vs Jupyter Classic vs nteract vs Arcadia, etc.

Reply:
Related to that, @colombod is adding an API for grabbing properties like this: #119.